crates/sui-core/src/authority.rs

```diff
@@ -4549,17 +4537,19 @@ impl AuthorityState {
             .execution_lock_for_executable_transaction(&executable_tx)
             .await?;
 
-        let input_objects = self
-            .input_loader
-            .read_objects_for_synchronous_execution(
-                &executable_tx
-                    .data()
-                    .intent_message()
-                    .value
-                    .input_objects()?,
+        // We must manually assign the shared object versions to the transaction before executing it.
+        // This is because we do not sequence end-of-epoch transactions through consensus.
+        epoch_store
+            .assign_shared_object_versions_idempotent(
+                self.get_cache_reader().as_ref(),
+                &[executable_tx.clone()],
             )
             .await?;
 
+        let input_objects = self
```

Comment:
I think we could try adding a crash fail point here to see if it is indeed idempotent. (Don't need to be in this PR).

Reply:
there are already fail points before/after db writes so we should already be able to catch any problems here

Reply:
Oh yeah, just saw `batch-write-before` and `batch-write-after` which is quite deep inside the stack. Nice!